pkg/autoscaler/stats_scraper.go

```diff
@@ -0,0 +1,142 @@
+	"time"
+
+	"github.com/knative/serving/pkg/apis/serving"
+	"github.com/prometheus/common/expfmt"
+	"go.uber.org/zap"
+)
+
+// StatsScraper defines the interface for collecting Revision metrics
+type StatsScraper interface {
+	// Scrape scrapes the Revision queue metric endpoint then sends it as a
+	// StatMessage to the given channel.
+	Scrape(statsCh chan<- *StatMessage)
+}
+
+// cacheDisabledClient is a http client with cache disabled. It is shared by
+// every goruntime for a revision scraper.
+var cacheDisabledClient = &http.Client{
+	Transport: &http.Transport{
+		// Do not use the cached connection
+		DisableKeepAlives: true,
+	},
+	Timeout: 3 * time.Second,
+}
+
+// ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
+// pod to be picked up to serve the request is decided by K8S.
+type ServiceScraper struct {
+	httpClient *http.Client
+	url        string
+	metric     *Metric
+	metricKey  string
+	logger     *zap.SugaredLogger
+}
+
+// CreateNewServiceScraper creates a new StatsScraper for the Revision which
+// the given Metric is responsible for.
+func CreateNewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (ServiceScraper, error) {
+	return createServiceScraperWithClient(metric, logger, cacheDisabledClient)
+}
+
+func createServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (ServiceScraper, error) {
+	revName := metric.Labels[serving.RevisionLabelKey]
+	if revName == "" {
+		return ServiceScraper{}, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
+	}
+
+	return ServiceScraper{
+		httpClient: httpClient,
+		url:        fmt.Sprintf("http://%s-service.%s:9090/metrics", revName, metric.Namespace),
+		metric:     metric,
+		metricKey:  NewMetricKey(metric.Namespace, metric.Name),
+		logger:     logger,
+	}, nil
+}
+
+// Scrape call the destination service then send it
+// to the given stats chanel
+func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
+	stat, err := s.scrapeViaURL()
+	if err != nil {
+		s.logger.Errorw("Failed to get metrics", zap.Error(err))
+		return
+	}
+
+	s.sendStatMessage(stat, statsCh)
+}
+
+func (s *ServiceScraper) scrapeViaURL() (Stat, error) {
+	req, err := http.NewRequest("GET", s.url, nil)
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return Stat{}, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return Stat{}, fmt.Errorf("GET request for URL %q returned HTTP status %v", s.url, resp.StatusCode)
+	}
+
+	return extractData(resp)
+}
+
+func extractData(resp *http.Response) (Stat, error) {
+	stat := Stat{}
+	var parser expfmt.TextParser
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		return stat, fmt.Errorf("Reading text format failed: %v", err)
+	}
+
+	now := time.Now()
+	stat.Time = &now
+
+	if metric, ok := metricFamilies["queue_average_concurrent_requests"]; ok {
+		for _, label := range metric.Metric[0].Label {
```

Comment:
`parser.TextToMetricFamilies` drops empty `MetricFamily` so it at least has one data point. However, to be safer, it's better to handle empty `Metric` list, in case there's change to `parser.TextToMetricFamilies`. Added the check.